Tool to make hadoop bam fragments readable

droazen opened the pull request:
Spark troubleshooting utility that converts a Hadoop bam fragment (eg., a part0000, part0001, etc. file produced by ReadsSparkSink) into a readable bam file by adding a header and a BGZF terminator.

droazen commented:
@lbergelson please review

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/ConvertHadoopBamFragmentToBam.java:
```diff
@@ -0,0 +1,88 @@
+
+/**
+ * Spark troubleshooting utility that converts a Hadoop bam fragment (eg., a part0000, part0001, etc. file
+ * produced by ReadsSparkSink) into a readable bam file by adding a header and a BGZF terminator.
```
This should be a `{@link ReadSparkSink}` so it doesn't get forgotten on a rename.

droazen replied:
Done

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/ConvertHadoopBamFragmentToBam.java:
```diff
@@ -0,0 +1,88 @@
+        new SAMTextHeaderCodec().encode(stringWriter, samFileHeader, true);
+        headerString = stringWriter.toString();
+
+        outputBinaryCodec.writeBytes("BAM\1".getBytes());
```
Is this a magic bam header string?  Is it defined somewhere in htsjdk?

droazen replied:
Yes. It's defined in htsjdk, but the definition is package protected, so we have to duplicate it here.

lbergelson replied:
Gross.  Can you extract it and give it a name at least?

droazen replied:
Done

lbergelson commented:
We should have a test for this.  Just take any part file, show that it isn't a valid bam, then run this and show that it is.

lbergelson commented:
@droazen back to you

lbergelson commented:
@droazen Actually, one more comment.  Can you factor out the action of this tool into it's own method so it could be invoked directly instead of only through the CommandLineProgram?

tomwhite commented on src/main/java/org/broadinstitute/hellbender/tools/ConvertHadoopBamFragmentToBam.java:
```diff
@@ -0,0 +1,88 @@
+
+        try ( FileOutputStream outStream = new FileOutputStream(outputBam) ) {
+            writeHeader(outStream, header);
+            final byte[] fragmentBytes = IOUtils.readFileIntoByteArray(hadoopFragment);
```
This shouldn't read into memory as it will fail with larger BAMs.

Could you use ReadsSparkSink#mergeBam? This would also work for HDFS files and multiple Hadoop fragments.

droazen replied:
Well, this utility was written specifically to identify a bug in `ReadsSparkSink#mergeBam`, which is why it doesn't call into it (and for the same reason, I think it should continue to be independent). I'll modify it so that it doesn't slurp the entire fragment into memory at once, though, to allow it to work with larger fragments.

droazen replied:
Done -- switched to using apache's `FileUtils.copyFile(file, stream)`

tomwhite commented on src/main/java/org/broadinstitute/hellbender/tools/ConvertHadoopBamFragmentToBam.java:
```diff
@@ -0,0 +1,88 @@
+        oneLineSummary = "Convert a hadoop bam fragment into a readable bam",
+        programGroup = TestSparkProgramGroup.class
+)
+public final class ConvertHadoopBamFragmentToBam extends CommandLineProgram {
```
The terms 'sharded' is used in ReadsSparkSink, so it would be good to be consistent.

droazen replied:
Done

droazen commented:
Extracted the logic of the tool into a utility method, and added unit/integration tests. All comments addressed.

droazen commented:
@lbergelson @tomwhite I noticed while addressing the code review comments for this branch that `ReadsSparkSink` saves the "shards" *with* a header for the `ReadsWriteFormat.SHARDED` case, and *without* a header for the `ReadsWriteFormat.SINGLE` case. The tool in this branch assumes that the shards are headerless -- should it be able to work with both kinds of shards? The shards with a header are actually readable as-is with `samtools`

droazen commented:
Assigning to @lbergelson for clarification on the question raised above

lbergelson commented:
@droazen I think it would ideally handle either.  Since there isn't a non-debugging case where you would end up with the shards created by the `SINGLE` case, it should probably handle the `SHARDED` case.  I'm not sure there's an easy way to tell if they're sharded or not, other than opening one and seeing if you can read a header.  Should we change that, special sub-extension or something?

droazen commented:
In the interest of getting this merged, I've addressed the remaining blocking issue via documentation and naming: tool is now named `ConvertHeaderlessHadoopBamShardToBam`, and the docs make it clear what kinds of Hadoop bam shards it should be used with, and which it shouldn't. Will merge once tests pass.
